src/tls13.c

```diff
@@ -8519,35 +8591,62 @@ static int SendTls13Certificate(WOLFSSL* ssl)
         /* Cert Req Ctx Len | Cert Req Ctx | Cert List Len | Cert Data Len */
         headerSz = OPAQUE8_LEN + certReqCtxLen + CERT_HEADER_SZ +
                    CERT_HEADER_SZ;
+        XMEMSET(extSz, 0, sizeof(extSz));
 
-        ret = TLSX_GetResponseSize(ssl, certificate, &extSz);
+        length = 0;
+        listSz = 0;
+
+        ret = TLSX_GetResponseSize(ssl, certificate, &extSz[0]);
         if (ret < 0)
             return ret;
 
         /* Create extensions' data if none already present. */
-        if (extSz > OPAQUE16_LEN && ssl->buffers.certExts == NULL) {
-            ret = AllocDer(&ssl->buffers.certExts, extSz, CERT_TYPE, ssl->heap);
+        if (extSz[0] > OPAQUE16_LEN && ssl->buffers.certExts[0] == NULL) {
+            ret = AllocDer(&ssl->buffers.certExts[0], extSz[0],
+                CERT_TYPE, ssl->heap);
             if (ret < 0)
                 return ret;
 
-            extSz = 0;
-            ret = TLSX_WriteResponse(ssl, ssl->buffers.certExts->buffer,
-                                                           certificate, &extSz);
+            extSz[0] = 0;
+            ret = TLSX_WriteResponse(ssl, ssl->buffers.certExts[0]->buffer,
+                certificate, &extSz[0]);
             if (ret < 0)
                 return ret;
         }
+        /* set empty extension as default */
+        for (extIdx = 1; extIdx < (1 + MAX_CERT_EXTENSIONS); extIdx++)
+            extSz[extIdx] = OPAQUE16_LEN;
+
+        /* certificate extension size */
+        totalextSz += extSz[0];
```

Comment:
Can we use `WriteCSRToBuffer` to write out the leaf cert too?

Reply:
It needs additional header write for the leaf cert. So, I would like to leave it as is and change the function name to WriteCSRToChainBuffer.

Reply:
All of the certs and extensions have the same format. I don't see why we couldn't just do it in `WriteCSRToChainBuffer`.
![image](https://github.com/user-attachments/assets/1c20f95a-b3c9-4298-a241-dcfb7dacd4b4)

Reply:
That is true. Thanks. Use WriteCSRToBuffer for the leaf cert

Reply:
That is not what I had in mind. This is the patch that I want to see:
```
diff --git a/src/tls13.c b/src/tls13.c
index e0fe00b3f..eecef8df4 100644
--- a/src/tls13.c
+++ b/src/tls13.c
@@ -8417,7 +8417,7 @@ static word32 NextCert(byte* data, word32 length, word32* idx)
  * returns   Total number of bytes written.
  */
 static word32 WriteCSRToBuffer(WOLFSSL* ssl, DerBuffer** certExts,
-                                word16* extSz,  word16 extSz_num, word16 offset)
+                                word16* extSz,  word16 extSz_num)
 {
     int    ret = 0;
     TLSX* ext;
@@ -8434,7 +8434,7 @@ static word32 WriteCSRToBuffer(WOLFSSL* ssl, DerBuffer** certExts,
 
     if (csr) {
         for (extIdx = 0; extIdx < (word16)(extSz_num); extIdx++) {
-            tmpSz = TLSX_CSR_GetSize_ex(csr, 0, extIdx + offset);
+            tmpSz = TLSX_CSR_GetSize_ex(csr, 0, extIdx);
 
             if (tmpSz > (OPAQUE8_LEN + OPAQUE24_LEN) &&
                 certExts[extIdx] == NULL) {
@@ -8455,8 +8455,7 @@ static word32 WriteCSRToBuffer(WOLFSSL* ssl, DerBuffer** certExts,
                             + HELLO_EXT_TYPE_SZ + OPAQUE16_LEN);
                 /* write extension data */
                 extSz[extIdx] = (word16)TLSX_CSR_Write_ex(csr,
-                        der->buffer + ex_offset, 0,
-                        extIdx + offset);
+                        der->buffer + ex_offset, 0, extIdx);
                 /* add extension offset */
                 extSz[extIdx] += (word16)ex_offset;
                 /* extension length */
@@ -8467,8 +8466,8 @@ static word32 WriteCSRToBuffer(WOLFSSL* ssl, DerBuffer** certExts,
         }
     }
     else {
-            /* chain cert empty extension size */
-            totalSz += OPAQUE16_LEN * extSz_num;
+        /* chain cert empty extension size */
+        totalSz += OPAQUE16_LEN * extSz_num;
     }
     return totalSz;
 }
@@ -8593,41 +8592,30 @@ static int SendTls13Certificate(WOLFSSL* ssl)
         /* Cert Req Ctx Len | Cert Req Ctx | Cert List Len | Cert Data Len */
         headerSz = OPAQUE8_LEN + certReqCtxLen + CERT_HEADER_SZ +
                    CERT_HEADER_SZ;
-        XMEMSET(extSz, 0, sizeof(extSz));
-
-        length = 0;
-        listSz = 0;
-
-        ret = TLSX_GetResponseSize(ssl, certificate, &extSz[0]);
-        if (ret < 0)
-            return ret;
-
-        /* Create extensions' data if none already present. */
-    #if defined(HAVE_CERTIFICATE_STATUS_REQUEST)
-        if (extSz[0] > OPAQUE16_LEN && ssl->buffers.certExts[0] == NULL) {
-            extSz[0] =  WriteCSRToBuffer(ssl, &ssl->buffers.certExts[0],
-                                    &extSz[0], 1/* own cert */, 0/* offset */);
-        }
-    #endif
         /* set empty extension as default */
-        for (extIdx = 1; extIdx < (word16)(1 + MAX_CERT_EXTENSIONS); extIdx++)
+        for (extIdx = 0; extIdx < XELEM_CNT(extSz); extIdx++)
             extSz[extIdx] = OPAQUE16_LEN;
 
-        /* certificate extension size */
-        totalextSz += extSz[0];
+        length = 0;
+        listSz = 0;
 
     #if defined(HAVE_CERTIFICATE_STATUS_REQUEST)
-        if (ssl->options.side == WOLFSSL_SERVER_END &&
-            ssl->buffers.certChainCnt > 0) {
-            /* write CSR to the buffer */
-            totalextSz += WriteCSRToBuffer(ssl, &ssl->buffers.certExts[1],
-                                    &extSz[1], ssl->buffers.certChainCnt, 1);
-
+        /* We only send CSR on the server side. On client side, the CSR data
+         * is populated with the server response. We would be sending the server
+         * its own stapling data. */
+        if (ssl->options.side == WOLFSSL_SERVER_END) {
+            ret = WriteCSRToBuffer(ssl, &ssl->buffers.certExts[0], &extSz[0],
+                    1 /* +1 for leaf */ + ssl->buffers.certChainCnt);
+            if (ret < 0)
+                return ret;
+            totalextSz += ret;
+            ret = 0; /* Clear to signal no error */
         }
-        else if (ssl->options.side == WOLFSSL_CLIENT_END &&
-                ssl->buffers.certChainCnt > 0)
+        else
     #endif
         {
+            /* Leaf cert empty extension size */
+            totalextSz += OPAQUE16_LEN;
             /* chain cert empty extension size */
             totalextSz += OPAQUE16_LEN * ssl->buffers.certChainCnt;
         }
```